velox/exec/PrefixSort.cpp

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "PrefixSort.h"
+
+namespace facebook::velox::exec {
+
+namespace {
+
+template <TypeKind typeKind>
+FOLLY_ALWAYS_INLINE uint32_t normalizeSize() {
+  VELOX_UNSUPPORTED("PrefixSort not support the type.");
+}
+
+template <>
+FOLLY_ALWAYS_INLINE uint32_t normalizeSize<TypeKind::BIGINT>() {
+  return sizeof(TypeTraits<TypeKind::BIGINT>::NativeType);
+}
+
+FOLLY_ALWAYS_INLINE uint32_t normalizeSize(const TypeKind& typeKind) {
+  return VELOX_DYNAMIC_TYPE_DISPATCH(normalizeSize, typeKind);
+}
+
+template <TypeKind typeKind>
+FOLLY_ALWAYS_INLINE void normalizeRowColumn(
+    const PrefixSortLayout& prefixSortLayout,
+    const uint32_t index,
+    const RowColumn& rowColumn,
+    char* FOLLY_NONNULL const row,
+    char* FOLLY_NONNULL const prefix) {
+  VELOX_UNSUPPORTED("PrefixSort not support the type.");
+}
+
+template <>
+FOLLY_ALWAYS_INLINE void normalizeRowColumn<TypeKind::BIGINT>(
+    const PrefixSortLayout& prefixSortLayout,
+    const uint32_t index,
+    const RowColumn& rowColumn,
+    char* FOLLY_NONNULL const row,
+    char* FOLLY_NONNULL const prefix) {
+  using T = TypeTraits<TypeKind::BIGINT>::NativeType;
+  // Store null as min/max value according compare flags.
+  if (RowContainer::isNullAt(row, rowColumn.nullByte(), rowColumn.nullMask())) {
+    CompareFlags compareFlags = prefixSortLayout.compareFlag(index);
+    PrefixSortEncoder::encode(
+        ((compareFlags.ascending && compareFlags.nullsFirst) ||
+         (!compareFlags.ascending && !compareFlags.nullsFirst))
+            ? std::numeric_limits<T>::min()
+            : std::numeric_limits<T>::max(),
+        prefix + prefixSortLayout.prefixOffset(index));
+  } else {
+    PrefixSortEncoder::encode(
+        *(reinterpret_cast<T*>(row + rowColumn.offset())),
+        prefix + prefixSortLayout.prefixOffset(index));
+  }
+  // Invert bits if desc
```

Comment:
Shouldn't this logic be part of PrefixSortEncoder::encode ? We could pass sorting flags to encode.

Reply:
I think it`s better to put it into extractRowToPrefix to avoid write it everywhere, like:
```
void PrefixSort::extractRowToPrefix(char* row, char* prefix) {
  // Extract row to normalized prefix by column.
  for (int32_t index = 0; index < sortLayout_.numNormalizedKeys(); index++) {
    VELOX_DYNAMIC_TYPE_DISPATCH(
        normalizeRowColumn,
        rowContainer_->keyTypes()[index]->kind(),
        sortLayout_,
        index,
        rowContainer_->columnAt(index),
        row,
        prefix);
    // Invert bits if desc
    if (!sortLayout_.compareFlag(index).ascending) {
      for (uint64_t s = 0; s < sizeof(T); s++) {
        *(prefix + sortLayout_.prefixOffset(index) + s) =
            ~*(prefix + sortLayout_.prefixOffset(index) + s);
      }
    }
  }
  // Set row address.
  *reinterpret_cast<char**>(prefix + sortLayout_.normalizedKeySize()) = row;
}
```

Reply:
"Invert bits if desc" - If this logic applies to all types, then encoder can have a helper function for it and call it for all types. Having encoding logic in a single unit-testable place makes it easier to understand what that logic is and allows to write simple unit tests to verify that the logic is correct.